### PR TITLE
Faster Mapping Parsing on Master and Datanode

### PR DESCRIPTION
**Pull request by original-brownbear**
A couple of straightforward mapper parsing improvements.
Save round-trips to `String` in many spots.
Don't serialize the mapping again in the document mapper constructor
when the mapping is known to be correct (free of redundant white-paces etc.)
because we received it from the master/CS. This also saves a bit of memory
from not having duplicate `CompressedXContent` on heap on data nodes (because we
use the instances straight from the cluster state and those are deduplicated).

relates #77466

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/compress/CompressedXContent.java [hunk 1]
```diff
@@ -135,6 +137,20 @@ public CompressedXContent(byte[] data) throws IOException {
         this(new BytesArray(data));
     }
 
+    public static CompressedXContent fromJSON(String json) throws IOException {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/common/compress/CompressedXContent.java:140]: Added this to normalize the whitespaces away when doing the string to `CompressedXContent` conversion. This saves quite a bit of parsing if requests come from external sources (Beats etc.) which may or may not use formats including some whitespaces.

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java [hunk 2]
```diff
@@ -398,7 +393,7 @@ private boolean assertSerialization(DocumentMapper mapper) {
         final CompressedXContent mappingSource = mapper.mappingSource();
         Mapping newMapping = parseMapping(mapper.type(), mappingSource);
         if (newMapping.toCompressedXContent().equals(mappingSource) == false) {
-            throw new IllegalStateException(
+            throw new AssertionError(
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:396]: This was technically swallowing assertion failures so I fixed it.

File: server/src/test/java/org/elasticsearch/cluster/metadata/MetadataMappingServiceTests.java [hunk 3]
```diff
@@ -62,7 +62,7 @@ public void testClusterStateIsNotChangedWithIdenticalMappings() throws Exception
         final MetadataMappingService mappingService = getInstanceFromNode(MetadataMappingService.class);
         final ClusterService clusterService = getInstanceFromNode(ClusterService.class);
         final PutMappingClusterStateUpdateRequest request = new PutMappingClusterStateUpdateRequest("""
-            { "properties" { "field": { "type": "text" }}}""");
+            { "properties": { "field": { "type": "text" }}}""");
```
Comment (original-brownbear) [inline server/src/test/java/org/elasticsearch/cluster/metadata/MetadataMappingServiceTests.java:65]: Obviously broken but previously this passed weirdly enough. I'm investigating why separately.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java [hunk 4]
```diff
@@ -33,7 +32,10 @@
  */
 public class MappingMetadata extends AbstractDiffable<MappingMetadata> {
 
-    public static final MappingMetadata EMPTY_MAPPINGS = new MappingMetadata("_doc", Collections.emptyMap());
+    public static final MappingMetadata EMPTY_MAPPINGS = new MappingMetadata(
+        MapperService.SINGLE_MAPPING_NAME,
+        Map.of(MapperService.SINGLE_MAPPING_NAME, Map.of())
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java:37]: I adjusted this to have consistency with empty mappings that were manually/externally (and would be wrapped with `_doc`) created when using this constant.

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java [hunk 5]
```diff
@@ -29,14 +29,16 @@ public static DocumentMapper createEmpty(MapperService mapperService) {
...
-        return new DocumentMapper(mapperService.documentParser(), mapping);
+        return new DocumentMapper(mapperService.documentParser(), mapping, mapping.toCompressedXContent());
     }
 
-    DocumentMapper(DocumentParser documentParser, Mapping mapping) {
+    DocumentMapper(DocumentParser documentParser, Mapping mapping, CompressedXContent source) {
         this.documentParser = documentParser;
         this.type = mapping.getRoot().name();
         this.mappingLookup = MappingLookup.fromMapping(mapping);
-        this.mappingSource = mapping.toCompressedXContent();
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java:39]: This was (and on master still is) about 50% of the time for a mapping update. Saving it on data nodes helps in a measurable way in our benchmarking when setting up thousands of indices as quickly as possible.

File: server/src/test/java/org/elasticsearch/cluster/metadata/MetadataMappingServiceTests.java (hunk 3 above)
Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/cluster/metadata/MetadataMappingServiceTests.java:65]: https://github.com/elastic/elasticsearch/pull/82408

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java [hunk 6]
```diff
@@ -1211,18 +1215,17 @@ static IndexMetadata buildIndexMetadata(
...
-        List<Map<String, Object>> mergedMappings = new ArrayList<>(1 + mappings.size());
-        mergedMappings.add(indexMode.getDefaultMapping());
-        mergedMappings.addAll(mappings);
-        for (Map<String, Object> mapping : mergedMappings) {
-            if (mapping.isEmpty() == false) {
-                mapperService.merge(MapperService.SINGLE_MAPPING_NAME, mapping, MergeReason.INDEX_TEMPLATE);
-            }
+        final CompressedXContent defaultMapping = indexMode.getDefaultMapping();
+        if (defaultMapping != null) {
+            mapperService.merge(MapperService.SINGLE_MAPPING_NAME, defaultMapping, MergeReason.INDEX_TEMPLATE);
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:1225]: I get thrown off here by seeing that you merge in the default mapping, which previously we did not. Why is that?

File: server/src/main/java/org/elasticsearch/common/compress/CompressedXContent.java (hunk 1 above)
Comment (javanna) [inline server/src/main/java/org/elasticsearch/common/compress/CompressedXContent.java:140]: Could we have javadoc for this new method?

**Review by javanna — COMMENTED** [review]
I left some initial comments, thanks for working on this, some of these aspects have also bugged me at times.

File: server/src/test/java/org/elasticsearch/cluster/metadata/MetadataMappingServiceTests.java (hunk 3 above)
Comment (javanna) [reply server/src/test/java/org/elasticsearch/cluster/metadata/MetadataMappingServiceTests.java:65]: thanks for catching ....

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java [hunk 7]
```diff
@@ -224,6 +222,10 @@ public DocumentParser documentParser() {
      * Parses the mappings (formatted as JSON) into a map
      */
     public static Map<String, Object> parseMapping(NamedXContentRegistry xContentRegistry, String mappingSource) throws IOException {
+        if ("{}".equals(mappingSource)) {
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:225]: Yea I think this is 100% safe. The parsing below is only parsing a map from JSON and it's a static method, so `{}` => empty map holds everytime.
And yes, this is for the relatively common case of seeing the default `{}` from requests.

File: server/src/main/java/org/elasticsearch/index/IndexMode.java [hunk 8]
```diff
@@ -66,8 +66,8 @@ public void validateTimestampFieldMapping(boolean isDataStream, MappingLookup ma
         }
 
         @Override
-        public Map<String, Object> getDefaultMapping() {
-            return Collections.emptyMap();
+        public CompressedXContent getDefaultMapping() {
+            return null;
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/index/IndexMode.java:70]: Jup there's only one caller :) `org.elasticsearch.cluster.metadata.MetadataCreateIndexService#updateIndexMappingsAndBuildSortOrder`

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java (hunk 6 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:1225]: We did merge it previously as well. The only difference is that I stopped needlessly collecting the mappings into a `List`. We'd just merge 2 lines later when iterating the mappings list below.

File: server/src/main/java/org/elasticsearch/common/compress/CompressedXContent.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/common/compress/CompressedXContent.java:147]: ++ added

**Comment by original-brownbear** [comment]
Thanks for taking a look Luca! Addressed all points (didn't do the rename though since it's quite noisy, let me know what you think there).

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:225]: Ok I was referring specifically to comparing strings, and wondering about whitespaces, line returns etc. I guess you already thought about that.

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java [hunk 9]
```diff
@@ -333,32 +335,25 @@ boolean assertNoUpdateRequired(final IndexMetadata newIndexMetadata) {
...
-            mergeAndApplyMappings(mappingMetadata.type(), mappingMetadata.source(), reason);
+            merge(mappingMetadata.type(), mappingMetadata.source(), reason);
         }
     }
 
     public DocumentMapper merge(String type, CompressedXContent mappingSource, MergeReason reason) {
-        return mergeAndApplyMappings(type, mappingSource, reason);
-    }
-
-    private DocumentMapper mergeAndApplyMappings(String mappingType, CompressedXContent mappingSource, MergeReason reason) {
```
[the remark this answers is not in the document]
Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:353]: I see, I think I introduced the different naming only for private methods, but the merge that we have now are all public hence I would leave them as-is.

**Review by javanna — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Thanks Luca!

**Pull request merged**